samples/bluetooth/broadcast_audio_sink/src/main.c

```diff
@@ -76,6 +79,96 @@ static uint32_t requested_bis_sync;
 static uint32_t bis_index_bitfield;
 static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
 
+#if defined(CONFIG_LIBLC3)
+
+#include "lc3.h"
+
+#define MAX_SAMPLE_RATE       16000
+#define MAX_FRAME_DURATION_US 10000
+#define MAX_NUM_SAMPLES       ((MAX_FRAME_DURATION_US * MAX_SAMPLE_RATE) / USEC_PER_SEC)
+
+static int16_t audio_buf[MAX_NUM_SAMPLES];
+static lc3_decoder_t lc3_decoder;
+static lc3_decoder_mem_16k_t lc3_decoder_mem;
+static int frames_per_sdu;
+
+static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
+{
+	int8_t ret;
+	int freq_hz;
+	int frame_duration_us;
+
+	printk("Enable: stream with codec %p\n", codec_cfg);
+
+	ret = bt_audio_codec_cfg_get_freq(codec_cfg);
+	if (ret > 0) {
+		freq_hz = bt_audio_codec_cfg_freq_to_freq_hz(ret);
+	} else {
+		printk("Error: Codec frequency not set, cannot start codec.");
+		return -1;
+	}
+
+	ret = bt_audio_codec_cfg_get_frame_dur(codec_cfg);
+	if (ret > 0) {
+		frame_duration_us = bt_audio_codec_cfg_frame_dur_to_frame_dur_us(ret);
+	} else {
+		printk("Error: Frame duration not set, cannot start codec.");
+		return ret;
+	}
+
+	frames_per_sdu = bt_audio_codec_cfg_get_frame_blocks_per_sdu(codec_cfg, true);
+
+	lc3_decoder = lc3_setup_decoder(frame_duration_us, freq_hz, 0, /* No resampling */
+					&lc3_decoder_mem);
+
+	if (lc3_decoder == NULL) {
+		printk("ERROR: Failed to setup LC3 decoder - wrong parameters?\n");
+		return -1;
+	}
+	return 0;
+}
+
+static void lc3_decode_handler(struct k_work *work)
+{
+	int err = 0;
+	int offset = 0;
+	uint8_t *buf_data;
+	int octets_per_frame;
+	struct broadcast_sink_stream *sink_stream = CONTAINER_OF(
+		k_work_delayable_from_work(work), struct broadcast_sink_stream, lc3_decode_work);
+
+	k_mutex_lock(&sink_stream->lc3_decoder_mutex, K_FOREVER);
```

Comment:
You should keep the lock shorter by taking a reference to `sink_stream->in_buf` in a local variable here, so that the RX thread won't be blocked, otherwise this will be as blocking as just doing the decoding in the `recv` callback.

So basically something like

```c
k_mutex_lock(&sink_stream->lc3_decoder_mutex, K_FOREVER);
if (sink_stream->in_buf == NULL) {
        buf = net_buf_ref(sink_stream->in_buf);
        net_buf_unref(sink_stream->in_buf);
        sink_stream->in_buf = NULL;
}
k_mutex_unlock(&sink_stream->lc3_decoder_mutex);
```